### ISSUE TITLE
Build: reducing query count, prefetching, or intermediate tables

### ISSUE THREAD

**Issue by agjohnson** [readthedocs/readthedocs.org#11631]
After a few different attempts, it seems like the query complexity added with prefetching causes queries to take much longer than without prefetching (however query _count_ is much much lower, 30 vs 120 queries).

For more background on the removal, see:
https://github.com/readthedocs/readthedocs.org/pull/11621

The most expensive queries here are getting the latest build (which I think can be external) and getting the latest, non-external, successful build.

Some options here are:

- Use caching after database query. Saves some lookups but initial is still expensive and savings inconsistent.
- Use materialized views to cache at the database level. Managing these views is harder and outside ORM
- Use an intermediate table for latest/latest successful build on Project. This requires updating the relationship after each build.

**Comment by humitos** [comment]
> Use an intermediate table for latest/latest successful build on Project. This requires updating the relationship after each build.

Do we need an intermediate table for this? Why not just `OneToOne` fields on the `Project` object?

**Comment by agjohnson** [comment]
Yup, it could be any relation, not necessarily an intermediate table.

**Comment by humitos** [comment]
I thought a little more about this and I realized that:

- We need to use an API call from the builder to update the DB object
- We can use [`on_success`](https://github.com/readthedocs/readthedocs.org/blob/698f6f081c9ebc68d886c576bd116ebbe27f3d1b/readthedocs/projects/tasks/builds.py#L651) and `on_failure` Celery handlers to update `lastest_successful_build` and `latest_builds` fields respectively
- It's not clear if we need these latest builds to be linked to the `Project` or to the `Version` objects 🤔

**Comment by agjohnson** [comment]
> It's not clear if we need these latest builds to be linked to the Project or to the Version objects

Good point, I think both if we want to speed up both pages. Project and version listing both show the latest build and use `has_good_build`.

**Comment by ericholscher** [comment]
Seems like a bunch of work was done here. Should we call this issue done? 

https://github.com/readthedocs/readthedocs.org/pull/12385, https://github.com/readthedocs/readthedocs.org/pull/12400, https://github.com/readthedocs/readthedocs.org/pull/12602, https://github.com/readthedocs/readthedocs.org/pull/12662.
